x-pack/plugins/fleet/server/services/epm/archive/storage.ts

```diff
@@ -138,3 +147,89 @@ export async function getAsset(opts: {
 
   return storedAsset;
 }
+export const getEsPackage = async (
+  pkgName: string,
+  pkgVersion: string,
+  packageAssets: SavedObjectsBulkGetObject[],
+  savedObjectsClient: SavedObjectsClientContract
+) => {
+  const pkgKey = pkgToPkgKey({ name: pkgName, version: pkgVersion });
+  const soRes = await savedObjectsClient.bulkGet<PackageAsset>(packageAssets);
+  const paths = soRes.saved_objects.map((asset) => asset.attributes.asset_path);
+
+  // create the packageInfo
+  // TODO: this is mostly copied from validtion.ts, but we should save packageInfo somewhere
+  // so we don't need to do this again as this was already done either in registry or through upload
```

Comment:
I'm still getting the full picture of what's going on, but wanted to remind we have get/setPackageInfo methods

https://github.com/elastic/kibana/blob/139f0a362ab22eea3f1f44877e591cbafdd157ec/x-pack/plugins/fleet/server/services/epm/archive/cache.ts#L49

Same trade-offs as elsewhere re: memory cache

Reply:
For packages that are not yet in storage (packages installed before storage was available) or in cache we would need to still create the package info.  Also in the case the package IS in storage but not in cache, we don't have the package info.  One thing I can do is check first if its in `packageInfoCache`, if its not, create it from storage as is being done here and add it to `packageInfoCache`.